README.md
CKEditor 4 AddOn für REDAXO
===========================

Der [CKEditor](http://ckeditor.com/) in Version 4 für REDAXO inkl. eines Beispielmoduls.

Features
--------

* CKEditor Full
* REDAXO Modul
* Editor Profile
* Eigene Plugins benutzen
* Smart Strip Funktion: filtert leere P's und mehrfach vorkommende BR's heraus
* Linkdialog: Unterstützung für Links über REDAXO Linkmap und Medienpool
* Imagedialog: Unterstützung für Links über REDAXO Medienpool
* Vereinfachter Tabellen- und Imagedialog
* Kurzhilfe für REDAXO Redakteure  
  
  
  
Eigene CKEditor Plugins benutzen
--------------------------------

Eigene Plugins werden einfach in dass project Addon dort unter `/install/ckeditor/plugins/` gelegt. Diese werden automatisch in das CKEditor Plugin Verzeichnis kopiert bei Install, Reinstall und Update des CKEditor AddOns und stehen so immer zur Verfügung. Im Profil muss man dann noch unter `extraPlugins` das Plugin mit auflisten (durch Komma getrennt wenn mehrere). Alles weitere sollte man der Doku des Plugins entnehmen... ;) 

Bitte beachten: Sollten nach einem Update die eigene Plugins nicht kompatibel sein mit der neusten CKEditor Version kann der Editor unter Umständen nicht mehr geladen werden. Abhilft schafft hier die betreffenden Plugins im Profil unter `extraPlugins` zu entfernen.

CKEditor in Modulen einsetzen
------------------------------

```html
<textarea class="ckeditor" data-ckeditor-profile="lite" name="REX_INPUT_VALUE[1]">REX_VALUE[1]</textarea>
```

* Die Textarea muss lediglich die CSS-Klasse `ckeditor` zugewiesen bekommen. 
* Desweiteren regelt man über `data-ckeditor-profile` das zu ladende Profil. 
* Wenn nötig kann man über `data-ckeditor-height` die Höhe überschreiben (wird sonst aus dem Profil genommen).

CKEditor in den Metainfos einsetzen
-----------------------------------

* In dem Feldattribute-Feld: `class="ckeditor" data-ckeditor-profile="lite"`
* Optional ebenfalls möglich: `data-ckeditor-height="150"`

CKEditor in yForm einsetzen
---------------------------

* Im Individuelle Attribute-Feld: `{ "class" : "ckeditor", "data-ckeditor-profile" : "lite" }`
* Weitere Attribute kommagetrennt möglich.

CKEditor in MForm einsetzen
---------------------------

```php
$mform->addTextareaField(1); 
$mform->setLabel('Text');
$mform->addAttribute('class', 'ckeditor');
$mform->addAttribute('data-ckeditor-profile', 'lite');
$mform->addAttribute('data-ckeditor-height', '200'); // optional
```
Auch andere Schreibweisen sind möglich, siehe hierzu die MForm Dokumentation.

CKEditor in MBlock einsetzen
----------------------------

```php
$mform->addTextAreaField("$id.0.textarea", array(
    'label'                 => 'Text',
    'class'                 => 'ckeditor',
    'data-ckeditor-profile' => 'lite'
));
```
Auch andere Schreibweisen sind möglich, siehe hierzu die MBlock Dokumentation.

Custom Styles hinzufügen
------------------------

Das Profil muss wie folgt ergänzt werden:

* Die __Styles__ Combobox zur Toolbar hinzufügen
* __stylesSet__ definieren, ersetzt das Besetehende das definiert ist in `/assets/addons/ckeditor/vendor/styles.js`
* Custom CSS per __contentsCss__ für den Editor hinzufügen

Hier ein Lite Profil mit Custom Styles:

```javascript
{
    height: 400,
    fillEmptyBlocks: false,
    forcePasteAsPlainText: true,
    entities: false,
    linkShowTargetTab: false,
    format_tags: 'p;h2;h3',
    removePlugins: 'elementspath',
    extraPlugins: 'rex_help',
    removeDialogTabs: 'link:advanced',
    disallowedContent: 'p{margin,margin-bottom,margin-left,margin-right,margin-top};img{border-style,border-width,margin,margin-bottom,margin-left,margin-right,margin-top};table{width,height}[align,border,cellpadding,cellspacing,summary];caption;',
    toolbar: [
        ['Format', 'Styles'],
        ['Bold', 'Italic'],
        ['NumberedList', 'BulletedList', '-', 'Outdent', 'Indent'],
        ['Link', 'Unlink', 'Anchor'],
        ['Table'],
        ['PasteText', 'PasteFromWord'],
        ['Maximize'],
        ['rex_help']
        // no comma after last entry!!!
    ],
    stylesSet: [
        { name: 'Grün', element: 'span', attributes: { 'class': 'green' } },
        { name: 'Blau', element: 'span', attributes: { 'class': 'blue' } }
        // no comma after last entry!!!
    ],    
    contentsCss: [CKEDITOR.basePath + 'contents.css', 
        '.green { background: lightgrey; color: green; }' +
        '.blue { background: lightgrey; color: blue; }'
        // no plus after last entry!!!
    ]
    // no comma after last entry!!!
}
```

Ausgabe nachträglich manipulieren
---------------------------------

```php
<?php
$html = <<< EOT
REX_VALUE[id=1 output=html]
EOT;

echo strtoupper($html);
?>
```

Bilder als Media Manager Urls umschreiben 
-----------------------------------------

Hilfreich wenn man Bilder nicht über einen REX_MEDIABUTTON[] sondern per Image Button aus dem Medienpool holt.

```php
<?php
$html = <<< EOT
REX_VALUE[id=1 output=html]
EOT;

echo '<div class="ckeditor-output">';
echo rex_ckeditor::replaceImageTags($html, 'my_media_type');
echo '</div>';
?>
```

```css
.ckeditor-output img { } 
.ckeditor-output img[style*="left"] { }
.ckeditor-output img[style*="right"] { }
```


Prüfen ob ein CKEditor Profil existiert
---------------------------------------

```php
if (rex_ckeditor::profileExists('lite')) {
	// profil "lite" existiert
}
```

Profil anlegen
---------------------------------------
Methode zur Erstellung eines neuen Profils, z.B. bei Installation eines Addons oder Moduls

```php
rex_ckeditor::insertProfile('profilname', $description, $jscode, '1');
```

Beispiel mit vorheriger Prüfung:

```php
// Ist CKeditor verfügbar
if (rex_addon::get('ckeditor')->isAvailable()) {

	// Prüfe ob das gewünschte Profil existiert
	if (!rex_ckeditor::profileExists('profilname')) {

		// Erstelle das Profil
		rex_ckeditor::insertProfile('profilname', $description, $jscode, '1');
	}
}
```


CKEditor Toolbar Buttons
------------------------

Hinweis: Mache Buttons sind nur verfügbar wenn die zugehörigen CKEditor Plugins installiert wurden.

* Source, Save, NewPage, DocProps, Preview, Print, Templates, document
* Cut, Copy, Paste, PasteText, PasteFromWord, Undo, Redo
* Find, Replace, SelectAll, Scayt
* Form, Checkbox, Radio, TextField, Textarea, Select, Button, ImageButton, HiddenField (benötigt das Forms Plugin!)
* Bold, Italic, Underline, Strike, Subscript, Superscript, RemoveFormat
* NumberedList, BulletedList, Outdent, Indent, Blockquote, CreateDiv, JustifyLeft, JustifyCenter, JustifyRight, JustifyBlock, BidiLtr, BidiRtl
* Link, Unlink, Anchor
* CreatePlaceholder, Image, Flash, Table, HorizontalRule, Smiley, SpecialChar, PageBreak, Iframe, InsertPre
* Styles, Format, Font, FontSize
* TextColor, BGColor
* UIColor, Maximize, ShowBlocks
* About

Links
-----

* CKEditor Toolbar Configurator: http://nightly.ckeditor.com/17-02-23-07-09/standard/samples/toolbarconfigurator/index.html
* CKEditor Addon für REDAXO konfigurieren: http://usysto.net/blog/redaxo_ckeditor_addon.php
* REDAXO Artikel im Frontend editieren mit dem CKEditor: http://usysto.net/blog/redaxo_frontend_edit_mit_ckeditor.php
* Alle CKEditor Config-Optionen: http://docs.ckeditor.com/#!/api/CKEDITOR.config
* CKEditor Best Practices: http://docs.ckeditor.com/#!/guide/dev_best_practices
* Content Filtering (ACF): http://docs.ckeditor.com/#!/guide/dev_acf

Hinweise
--------

* Getestet mit REDAXO 5.3 - 5.14.3
* AddOn-Ordner lautet: `ckeditor`

Changelog
---------

siehe `CHANGELOG.md` des AddOns

Lizenz
------

* CKEditor AddOn: MIT-Lizenz, siehe `LICENSE.md` des AddOns
* CKEditor: siehe `LICENSE.md` des CKEditors

Credits
-------

* CKSource, pschuchmann, iceman-fx, Xong, webghostx, phoebusryan, skerbis, georgkaser, fietstouring, dergel, prenzlweb, krugar, cukabeka, IngoWinter, JeGr, alexplusde, RexDude



CHANGELOG.md
CKEditor AddOn - Changelog
==========================

### Version 4.20.2 - 03.03.2023

* Update auf CKEditor v4.20.2 Full
* Fixed: Linkmap öffnete sich immer nur in der ersten Sprache

### Version 4.16.2 - 12. August 2021

* Update auf CKEditor v4.16.2 Full
* Fixed: Einbindung multipler CKEditoren in MBlock / Gridblock

### Version 4.16.1 - 20. Mai 2021

* Update auf CKEditor v4.16.1 Full
* Fixed: Neusortierung von MBlock in Verbindung mit ckeditor.
* Stylesheetparser bei Plugins hinzugefügt

### Version 4.13.1 - 21. Februar 2020

* Update auf CKEditor v4.13.1 Full
* Der CKEditor friert nun nicht mehr ein bei der Neusortierung in MBlock.

### Version 4.13.0 - 16. Oktober 2019

* Update auf CKEditor v4.13.0 Full
* Der CKEditor ist nun kompatibel mit MBlock.

### Version 4.11.3.1 - 19. März 2019

* Fixed: Der CKEditor wurde nicht vollständig geladen.

### Version 4.11.3 - 13. März 2019

* Update auf CKEditor v4.11.3 Full
* Smart Strip enfernt nun auch leere P-Tags die ein style Attribute haben (z.B. bei zentriertem Text), thx@KlemensWinkler

### Version 4.11.2 - 06. Februar 2019

* Update auf CKEditor v4.11.2 Full
* Mblock 3.x Full Compat, thx@fietstouring, thx@joachimdoerr

### Version 4.11.1 - 07. Januar 2019

* Update auf CKEditor v4.11.1 Full
* Mblock 3.x teilweise Compat, thx@fietstouring
* `rex_ckeditor::getProfiles()` hinzugfügt, thx@TobiasKrais

### Version 4.9.2 - 09. Juni 2018

* Update auf CKEditor v4.9.2 Full
* Ausgabe der Codebeispiele für REDAXO 5.6 wiederhergestellt
* Spanische Übersetzung hinzugefügt, thx@nandes2062
* disallowedContent ergänzt da sich margin Werte in den p-Tags verirren konnten
* Profile Cache-File wird beim REDAXO Cache löschen und beim REDAXO DB Import neu erzeugt

### Version 4.8.0-beta1 - 02. November 2017

* Update auf CKEditor v4.8.0 Full Nightly vom 02.11.2017 wegen Styleset Probleme in CKEditor v4.7.2, thx@rkemmere
* `rex_ckeditor::insertProfile()` ergänzt, thx@skerbis

### Version 4.7.2 - 25. August 2017

* Update auf CKEditor 4.7.2 Full
* Syntax Check vor Speicherung des JavaScript Config Objekts beim Profil hinzufügen/bearbeiten
* Methode `rex_ckeditor::profileExists()` hinzugefügt, prüft ob ein Profil vorhanden ist, thx@rkemmere
* Der CKEditor lässt sich nun auch über die CSS Klasse `rex-ckeditor` ansprechen. Ein Geschwindigkeitsvorteil bei der Editor-Initialisierung ist dadurch möglich (aber noch nicht wirklich bewiesen)..., thx@joachimdoerr
* Readme: "CKEditor in MBlock einsetzen" Abschnitt hinzugefügt

### Version 4.6.2 - 15. März 2017

* Portierung zu REDAXO 5
* Update auf CKEditor 4.6.2 FULL!
* Vereinfachter Initialisierungs-Code ohne JavaScript, siehe Readme
* Neuer Readme Eintrag: Bilder als Media Manager Urls umschreiben
* Extra Plugins Mechanismus eingebaut, siehe Readme
* Editor Profile
* Smart Strip Einstellung ins Profil verschoben
* Ab jetzt: AddOn Version gleich mit CKEditor Version

### Version 3.1.1 - 14. März 2016

* Update: Bitte die Hinweise in der `UPDATE.md` beachten!
* Medienpool Link-Button fixed. Bei Einbindung des CKEditors als Metainfo im Medienpool ist nur eine manuelle Linkeingabe ohne Link-Buttons möglich.

### Version 3.1.0 - 01. März 2016

* Update: Bitte die Hinweise in der `UPDATE.md` beachten!
* Update auf CKEditor 4.5.7
* Fixed #22: CKEditor sollte nun auch im Medienpool funktionieren

### Version 3.0.0 - 10. August 2015

* Update: Bitte die Hinweise in der `UPDATE.md` beachten!
* Update auf CKEditor 4.5.2
* Toolbar Configurator unter Hilfe > Tools hinzugefügt

### Version 2.6.0 - 01. Februar 2015

* Update: Bitte die Hinweise in der `UPDATE.md` beachten!
* Update auf CKEditor 4.4.7
* Advanced Tab im Link Dialog wird beim Standard Modul per default angezeigt
* Update Meldung (wenn Update per Installer durchgeführt) + `UPDATE.md` hinzugefügt (auch über Hilfe verfügbar)

### Version 2.5.1 - 03. Dezember 2014

* Module auf eine eigene Seite verschoben

### Version 2.5.0 - 25. Oktober 2014

* Updatefähigkeit für REDAXO 4.6 hergestellt. Einstellungen werden jetzt im Data-Ordner gespeichert.

### Version 2.4.0 - 25. September 2014

* Update auf CKEditor 4.4.5
* Fixed #19: Das deaktivieren der Smart Strip Einstellung führte dazu das Linkmap und Co. nicht mehr verfügbar waren im Editor, thx@krugar
* Optimierungen am Image Modul vorgenommen 
* Hinweis zum Addon-Update in die Readme aufgenommen: Bei einem Update des Addons sollte nach der Reinstallation immer der Browser-Cache komplett gelöscht werden!

### Version 2.3.0 - 29. Juli 2014

* Update auf CKEditor 4.4.3
* Neues Image Modul (Text + Bild) hinzugefügt. Benötigt das be_utilities > jquery_ui Plugin
* XForm Table Manager CSS verbessert

### Version 2.2.1 - 17. April 2014

* Abschnitt "Mehrere Editoren mit gleicher Konfiguration" zur Readme hinzugefügt
* Abschnitt "Ausgabe nachträglich manipulieren" zur Readme hinzugefügt
* Kleinere Readme-Korrekturen

### Version 2.2.0 - 08. Februar 2014

* Fixed #17: Geschützte Leerzeichen werden nicht mehr komplett entfernt, thx@webghost
* Fixed #16: Smart Strip Einstellung greift jetzt auch, wenn man den CKEditor per CSS Klasse `ckeditor` aufruft
* Resize Grip Einstellung hinzugefügt um zu steuern ob der Resize Grip zum vergrößern des Editors angezeigt wird
* Weitereführende Links und Artikel u.a. von webghost in die Readme gepackt

### Version 2.1.1 - 27. Januar 2014

* Image2 Plugin aus Litemodul entfernt, da unter Umständen Bilder per Copy And Paste im Editor landen konnten
* Fehlende ColorButtons zum Standardmodul hinzugefügt
* Bestätigungsdialog bei Modulaktualisierung hinzugefügt
* Finetuning

### Version 2.1.0 - 24. Januar 2014

* Module verbessert, CKEditor Replace jetzt per Textarea ID
* Custom Styles Beispiel in der Readme verbessert (vollständiges Beispiel-Modul hinzugefügt). Kommt jetzt ganz ohne eine extra CSS-Datei aus, thx@webghost
* Bilder haben einen Abstand bekommen zum Text
* Standard Modul: Alle verfügbaren Buttons und somit Funktionen hinzugefügt

### Version 2.0.0 - 23. Januar 2014


assets/redaxo.js
CKEDITOR.replaceClass = '';


//CKEDITOR LINKMAP/MEDIAPOOL CONNECT
function rex_ckeditor_get_link_from_linkmap() {
	var clang = 1;
		var urlQuery = new URLSearchParams(window.location.search);
		if (urlQuery.has('clang')) {
			clang = urlQuery.get('clang');
		}

	var linkMap = openLinkMap('', '&clang='+clang);

	$(linkMap).on('rex:selectLink', function (e, linkurl, linktext) {
		e.preventDefault();
		linkMap.close();
		
		jQuery('.rex-url input').val(linkurl);
		jQuery('.rex-protocol option:last').prop('selected', true);
	});
}

function rex_ckeditor_get_link_from_mediapool() {
  var mediapool = openMediaPool('ckeditor_medialink');

  $(mediapool).on('rex:selectMedia', function (e, filename) {
    e.preventDefault();
    mediapool.close();

    jQuery('.rex-url input').val("/media/" + filename);
    jQuery('.rex-protocol option:last').prop('selected', true); // only for link dialog
  });
}


//CKEDITOR MBLOCK / GRIDBLOCK COMPAT
$(document).on('rex:ready', function (e, container) {

	container.find('.ckeditor').each(function(){
		
		//init only new editor instances
		ckid = $(this).attr('id');
		instance = CKEDITOR.instances[ckid];

		if (!instance) {
			rex_ckeditor_init($(this));
		}
	});
  
  
	// dialog changes
	CKEDITOR.on('dialogDefinition', function(ev) {
	var dialogName 	= ev.data.name;
	var dialogTabs 	= ev.data.definition;
	
	
	// Scroll-Up Fix
	dialogObj	= ev.data.definition.dialog;
		dialogObj.on('show', function(){ jQuery('html').addClass('cke_dialog_open'); });
		dialogObj.on('hide', function(){ jQuery('html').removeClass('cke_dialog_open'); });
	

    // Plugin image ///////////////////////////////////////////////////////
    if (dialogName == 'image') {
      var infoTab = dialogTabs.getContents('info');
      var urlField = infoTab.get('txtUrl');
      urlField['className'] = 'rex-url';

      infoTab.remove('htmlPreview');
      infoTab.remove('medialink');
      infoTab.add({
        type: 'button',
        id: 'medialink',
        label: 'Medienpool Link',
        align: 'center',
        style: 'display:inline-block; position: absolute; right: 23px; top: 116px;',
        onClick: function () {
          rex_ckeditor_get_link_from_mediapool();
        }
      });
    } //endif

    // Plugin link /////////////////////////////////////////////////////////
    if (dialogName == 'link') {

      // auf das Tab 'info' bezogen
      var infoTab = dialogTabs.getContents('info');
      // Entfernen da nicht kompatibel mit Email Obfuscator
      infoTab.remove('emailSubject');
      infoTab.remove('emailBody');
      var linkType = infoTab.get('linkType');
      linkType['items'] = [ // unwichtig - nur Reihenfolge geändert
        ['URL', 'url'], //###lang
        ['E-Mail', 'email'], //###lang
        ['Anker in diesem Block', 'anchor'] //###lang
      ];
      var protocol = infoTab.get('protocol');
      protocol['className'] = 'rex-protocol';
      protocol['items'] = [
        ['http://\u200E', 'http://'],
        ['https://\u200E', 'https://'],
        ['<andere>', ''] //###lang
      ];
      var url = infoTab.get('url');
      url['className'] = 'rex-url';
      url['onKeyUp'] = function () {
        this.allowOnChange = false;
        var protocolCmb = this.getDialog().getContentElement('info', 'protocol'),
          url = this.getValue(),
          urlOnChangeProtocol = /^(http|https):\/\/(?=.)/i,
          urlOnChangeTestOther = /^((redaxo|files)|[#\/\.\?])/i;
        var protocol = urlOnChangeProtocol.exec(url);
        if (protocol) {
          this.setValue(url.substr(protocol[0].length));
          protocolCmb.setValue(protocol[0].toLowerCase());
        } else if (urlOnChangeTestOther.test(url))
          protocolCmb.setValue('');
        this.allowOnChange = true;
      };
      infoTab.remove('urlOptions');
      infoTab.add({
        type: 'vbox',
        id: 'urlOptions',
        children: [{
          type: 'hbox',
          widths: ['25%', '75%'],
          children: [{
            id: 'protocol',
            type: 'select',
            className: 'rex-protocol',
            label: 'Protokoll', //###lang
            'default': 'http://',
            items: [
              // Force 'ltr' for protocol names in BIDI. (#5433)
              ['http://\u200E', 'http://'],
              ['https://\u200E', 'https://'],
              ['<andere>', ''] //###lang
            ],
            setup: function (data) {
              if (data.url)
                this.setValue(data.url.protocol || '');
            },
            commit: function (data) {
              if (!data.url)
                data.url = {};

              data.url.protocol = this.getValue();
            }
          }, {
            type: 'text',
            id: 'url',
            label: 'URL', //###lang
            required: true,
            className: 'rex-url',
            onLoad: function () {
              this.allowOnChange = true;
            },
            onKeyUp: function () {
              this.allowOnChange = false;
              var protocolCmb = this.getDialog().getContentElement('info', 'protocol'),
                url = this.getValue(),
                urlOnChangeProtocol = /^(http|https):\/\/(?=.)/i,
                urlOnChangeTestOther = /^((redaxo|files)|[#\/\.\?])/i;

              var protocol = urlOnChangeProtocol.exec(url);
              if (protocol) {
                this.setValue(url.substr(protocol[0].length));
                protocolCmb.setValue(protocol[0].toLowerCase());
              } else if (urlOnChangeTestOther.test(url))
                protocolCmb.setValue('');

              this.allowOnChange = true;
            },
            onChange: function () {
              if (this.allowOnChange) // Dont't call on dialog load.
                this.onKeyUp();
            },
            validate: function () {
              var dialog = this.getDialog();

              if (dialog.getContentElement('info', 'linkType') && dialog.getValueOf('info', 'linkType') != 'url')
                return true;

              if ((/javascript\:/).test(this.getValue())) {
                alert('ungültiger Wert'); //###lang
                return false;
              }

              if (this.getDialog().fakeObj) // Edit Anchor.
                return true;

              var func = CKEDITOR.dialog.validate.notEmpty('URL fehlt'); //###lang
              return func.apply(this);
            },
            setup: function (data) {
              this.allowOnChange = false;
              if (data.url)
                this.setValue(data.url.url);
              this.allowOnChange = true;

            },
            commit: function (data) {
              // IE will not trigger the onChange event if the mouse has been used
              // to carry all the operations #4724
              this.onChange();

              if (!data.url)
                data.url = {};

              data.url.url = this.getValue();
              this.allowOnChange = false;
            }
          }],
          setup: function (data) {
            if (!this.getDialog().getContentElement('info', 'linkType'))
              this.getElement().show();
          }
        }, {
          type: 'button',
          id: 'internallink',
          label: 'Interner Link',
          style: 'float : right;',
          onClick: function () {
            rex_ckeditor_get_link_from_linkmap();

          }
        }, {
          type: 'button',
          id: 'medialink',
          label: 'Medienpool Link',
          style: 'float : right;',
          onClick: function () {
            rex_ckeditor_get_link_from_mediapool();
          }
        }]
      });
      // auf das Tab 'target' bezogen
      var targetTab = dialogTabs.getContents('target');
      var linkTargetType = targetTab.get('linkTargetType');
      linkTargetType['items'] = [
        ['normal', 'notSet'],
        ['neues Fenster', '_blank']
      ];
    } //endif
  }); // end dialogDefinition
});

function rex_ckeditor_init(element) {
  if (!element.next().hasClass('ckeditor')) {

    let unique_id = 'ckeditor' + Math.random().toString(16).slice(2);
    element.attr('id', unique_id);
	
	
    // set config object
    if (element.attr('data-ckeditor-profile') && element.attr('data-ckeditor-profile') in ckProfiles) {
      currentProfileName = element.attr('data-ckeditor-profile');
    } else {
      currentProfileName = ckDefaultProfileName;
    }

    currentEditorConfig = ckProfiles[currentProfileName];

    // add smart strip class
    if (ckSmartStripSettings[currentProfileName] === 1) {
      element.addClass('ckeditor-smartstrip');
    }

    // overwrite height if necessary
    if (element.attr('data-ckeditor-height')) {
      currentEditorConfig.height = element.attr('data-ckeditor-height');
    }

    // make sure rex_help is available otherwise js error       
    if (ckRexHelpPluginAvailable) {
      currentEditorConfig.extraPlugins = currentEditorConfig.extraPlugins + ',rex_help';
    } else {
      currentEditorConfig.extraPlugins = currentEditorConfig.extraPlugins.replace(/rex_help/g, '');
    }
	

    // init editor
    editor = document.getElementById(unique_id);
    CKEDITOR.replace(editor, currentEditorConfig);
	

    // smart strip/
    $('form').submit(function () {
      if ($('.ckeditor-smartstrip').length) {
        var data = CKEDITOR.instances[unique_id].getData();
        var doDataUpdate = false;

        // replace &nbsp;
        if (data.indexOf("&nbsp;") != -1) {
          data = data.replace(/&nbsp;/g, "  ");
          doDataUpdate = true;
        }

        // replace multiple <br>s with a single one
        if (data.indexOf("<br />\n<br />") != -1 || data.indexOf("<br />\r\n<br />") != -1 || data.indexOf("<br /><br />") != -1) {
          data = data.replace(/(<br\s*\/?>\s*)+/igm, "<br />\n");
          doDataUpdate = true;
        }

        // replace leading <br>s
        if (data.match(/(<(?!br)(\w)[^>]*>)(\s*<br\s*\/?>\s*)+/igm)) {
          data = data.replace(/(<(?!br)(\w)[^>]*>)(\s*<br\s*\/?>\s*)+/igm, '$1');
          doDataUpdate = true;
        }

        // replace trailing <br>s
        if (data.match(/(\s*<br\s*\/?>\s*)+(<\/(?!br)(\w)>)/igm)) {
          data = data.replace(/(\s*<br\s*\/?>\s*)+(<\/(?!br)(\w)>)/igm, '$2');
          doDataUpdate = true;
        }

        // replace empty paragraphs
        if (data.match(/(<p[^>]*>(\s|&nbsp;|<br\s*\/?>)*<\/p>)/igm)) {
          data = data.replace(/(<p[^>]*>(\s|&nbsp;|<br\s*\/?>)*<\/p>)/igm, '');
          doDataUpdate = true;
        }

        if (doDataUpdate) {
          CKEDITOR.instances[unique_id].setData(data);
        }

        return true;
      }

    });
  }
}

$(document).on('rex:change', function (e, container) {	
 	container.parent('div').find('textarea.ckeditor').each(function(){
 		let ed = $(this).next('.cke');		
 		if (ed.length > 0) {
 			let id = $(ed).attr('id').replace('cke_', '');	
 						
 			CKEDITOR.instances[id].destroy();
 			rex_ckeditor_init($(this));
 		}
 	})
});